src/main/scala/lms/thirdparty/nccl/nccl.scala

```diff
@@ -307,11 +307,72 @@ trait NCCLOps extends CLibs with SizeTOps with CudaOps {
    * NCCL primitives are generally not thread-safe, however, they are reentrant. Multiple threads should use separate communicator objects.
    */
 
+
+  /**
+   * Sendrecv
+   * Exchange data between two ranks, both sending and receiving at the same time
+   */
+  def ncclSendRecv[T:Manifest](sendbuff: Rep[Array[T]], recvbuff: Rep[Array[T]], count: Rep[Int], datatype: Rep[ncclDataTypeT], peer: Rep[Int],
+      comm: Rep[ncclCommT], stream: Rep[cudaStreamT]) = {
+    ncclCheck(ncclGroupStart())
+    ncclCheck(ncclSend(sendbuff, count, datatype, peer, comm, stream))
+    ncclCheck(ncclRecv(recvbuff, count, datatype, peer, comm, stream))
+    ncclCheck(ncclGroupEnd())
+  }
+
+  /**
+   * One-to-all (scatter)
+   * A one-to-all operation from a root rank can be expressed by merging all send and receive operations in a group
+   * `root` is the root rank, `myRank` is the rank of current device, `nRanks` is the number of devices associated with
+   * the communicator `comm`.
+   */
+  def ncclScatter[T:Manifest](sendbuff: Rep[Array[Array[T]]], recvbuff: Rep[Array[T]], root: Rep[Int], myRank: Rep[Int], nRanks: Rep[Int],
+      count: Rep[Int], datatype: Rep[ncclDataTypeT], comm: Rep[ncclCommT], stream: Rep[cudaStreamT])(implicit __pos: SourceContext) = {
+    ncclCheck(ncclGroupStart())
+    __ifThenElse(equals(myRank, root), {
```

Comment:
Doesn't `if` work here?

Reply:
no. We cannot use @virtualize here unfortunately.